Add new UI for additional source columns and remove feature flag

melissa-barca opened the pull request:
This PR remove the user pref `enable_additional_columns` so the feature will be available without specifying the option. Global Options Pane Layout will now look like the following screenshots. Users are able to add and remove source columns so they can look at multiple editors at once. 

This PR also includes some adjustments to the `consolidateColumns` function. This function previously didn't work as expected but wasn't being used so caused no issues. 

There is still a fair amount of work to do for this feature, this is just the next big step. 

Updated screenshot 7/14:
![image](https://user-images.githubusercontent.com/5323711/87462868-70ce0300-c5c5-11ea-8d78-9559d73d57f7.png)

![image](https://user-images.githubusercontent.com/5323711/87462895-7af00180-c5c5-11ea-93f7-e827e8002c3c.png)

jmcphers commented on src/cpp/session/resources/schema/user-prefs-schema.json:
```diff
@@ -210,12 +210,6 @@
         },
-        "enable_additional_columns": {
-           "type": "boolean",
-           "default": false,
```
What do you think about leaving this option in and just changing the default to `true`? The reason I suggest this is that it's come up at least once in conversation that in some environments we might actually want to constrain the UI (e.g. in a tutorial mode where we need to preserve the "only one file is open" invariant). We could use the option to do this since you've already done the wiring, though we might want to rename it to something more specific.

melissa-barca replied:
Done! I added "Allow Source Columns" which defaults to true.

gtritchie replied:
Anything to worry about if a user has added source columns, then switches off this preference? Do existing source columns get closed or do they continue to function, that sort of thing?

melissa-barca replied:
The columns will persist until the user refreshes the browser, then any open tabs are moved to the main source column. The `Remove Column` button in Global Options is still enabled after the option is turned off so the user could go there to remove columns. Removing them as soon as the option is set to Off seems jarring.

gtritchie replied:
Not suggesting we remove them when option is turned off, was just curious how it behaves.

jmcphers commented on src/gwt/src/org/rstudio/studio/client/workbench/prefs/views/PreferencesDialog.css:
```diff
@@ -27,16 +27,28 @@
    vertical-align: top;
-   padding-top: 0px;
+   padding: 0px;
+   border: 1px solid #1e77bc;
```
Maybe personal preference but I think it looks a little odd to have square corners (from the options dropdowns) right up against these rounded ones. Can we tone these borders down a little (just a dark gray) and reduce the radius?

melissa-barca replied:
I changed the border and added a screenshot to this PR's description.

gtritchie commented on src/gwt/src/org/rstudio/studio/client/workbench/prefs/views/PaneLayoutPreferencesPane.java:
```diff
@@ -320,13 +331,121 @@ else if (!otherIndices.get(i))
+         grid_.addStyleName(res_.styles().paneLayoutTable());
+         grid_.setCellSpacing(GRID_CELL_SPACING);
+         grid_.setCellPadding(GRID_CELL_PADDING);
+
```
Recommend giving this table an accessible name; this is a cheap but effective accessibility improvement for screen reader users.

         Roles.getGridRole().setAriaLabelProperty(grid_.getElement(), "Columns and Panes Layout");

kevinushey commented:
Some drive-by comments:

Could we tweak the display so that "Source" is displayed at the top of the source column (like a title), rather than in the center? A few more pixels of padding would also help the display, I think.

Would it be worthwhile to adjust the visibility of the Remove command based on whether there are any source columns present? (ie: gray it out when there's nothing to remove)

Should we consider restricting the maximum number of source columns as well? (does the IDE / UI fall apart if someone clicks Add Column too many times?)

melissa-barca commented:
I've updated the original post with a new screenshot. 'Source' labels are now at the top of the column, added padding to the select drop downs, and updated the button click handlers to be enabled or disabled when appropriate.
When I'd originally added 'Source' to the top of columns I didn't like the look of the mostly empty column but it does look better with the padding.

melissa-barca commented:
Regarding the max number of columns, for this PR I chose to limit it to 3 because after that things start getting ugly in this options pane. I think there should be a limit to the number of columns a user can have, but not so sure this is the best reasoning.

gtritchie commented:
> Regarding the max number of columns, for this PR I chose to limit it to 3 because after that things start getting ugly in this options pane

This seems reasonable, though no doubt somebody with a crazy-wide monitor and young eyesight might want more. I suppose you could create a preference for that (defaulting to 3). Don't think it needs to be exposed in the Options dialog, but having it there gives a quick answer anybody who wants it. But then we have to test that, and like you say the pane will start going wonky, so probably tidier to just leave hardcoded IMHO...
